File: java/datastructures/lib/src/main/java/datastructures/queue/Queue.java
package datastructures.queue;

import datastructures.linkedlist.Node;

public class Queue {

  private Node<Integer> rear, front;

  //NOTE: head = rear, tail = front
  public Node getRear()
  {
    return rear;
  }
  public Queue()
  {
    rear = null;
    front = null;
  }
  public void enqueue(int data)
  {
    if(rear == null)
    {
      rear = new Node(data);
      front = rear;
      return;
    }
    Node current = rear;
    while(current != null)
    {
      if(current.getNext() == null)
      {
        Node temp = new Node(data);
        temp.setNext(current.getNext());
        current.setNext(temp);
        front = temp;
        //temp = rear;
        break;
      }
      current = current.getNext();
    }
    return;

  }

  public int dequeue() throws NullPointerException
  {
    if(rear == null)
    {
      throw new NullPointerException("Queue is empty");
    }
    int x = rear.getData();
    rear = rear.getNext();
    if(rear == null)
    {
      front = rear;
    }
    return x;
  }

  public int peek() throws NullPointerException
  {
      if(rear == null)
        throw new NullPointerException("Queue is empty");
      return front.getData();
    }

  public boolean isEmpty() {
    return rear == null? true:false;
  }

  @Override
  public String toString()
  {
    String toDisplay = "";
    Node current = rear;
    while(current != null){
      toDisplay = toDisplay + "{" + current.getData() + "} -> ";
      current = current.getNext();
    }
    return toDisplay + "NULL";
  }

}

File: python/data_structures/queue.py
from data_structures.invalid_operation_error import InvalidOperationError

class _QueueNode:

    def __init__(self, value):
        self.value = value
        self.next = None



class Queue:

    def __init__(self):
        self.front = None
        self.rear = None
    
    def enqueue(self, value):
        node = _QueueNode(value)
        if self.rear is None:
            self.front = self.rear = node
        else:
            self.rear.next = node 
            self.rear = node

    def dequeue(self):
        if self.front is None:
            raise InvalidOperationError("Queue is empty")

        value = self.front.value
        self.front = self.front.next
        if self.rear is None:
            self.rear = None
        return value
    
    def peek (self):
        if self.front is None:
            raise InvalidOperationError("Queue is empty")
        return self.front.value
    
    def is_empty(self):
        return self.front is None

    


